[CMake] Fix several issues that caused unnecessary rebuild

rintaro opened the pull request:
`build-script --reconfigure` should be "no work to do" unless anything has changed.

* [cxx-interop] Use add_custom_command_target for copying std.apinotes
`COMMAND` for `add_custom_target` are ["always considered out of date"](https://cmake.org/cmake/help/latest/command/add_custom_target.html) and executed every time. Because of that, targets depends on that were always marked "dirty". Use `add_custom_command_target` instead.
* [stdlib]  Copy '.private.swiftinteface' to swift_static
`.private.swiftinterface` should be a part of the module. And it's declared as `OUTPUT` file list, but they were never created. Because of that the target was always rebuilt.
* [stdlib] Update source file list only if different
Write source file list using `file(GENERATE` so that they are updated only if thier content is changed. Otherwise stdlib modules are rebuild every time cmake configuration happens.

rintaro committed:
[CMake] Use add_custom_command_target for copying std.apinotes

`COMMAND` for `add_custom_target` are "always considered out of date"
and executed everytime. Because of that, targets depends on that were
alway marked "dirty". Use `add_custom_command_target` instead.

rintaro commented:
@swift-ci Please smoke test

egorzhdan reviewed (approved):
Thank you!

rintaro commented on stdlib/public/Cxx/std/CMakeLists.txt:
```diff
@@ -2,7 +2,9 @@
-add_custom_target(CxxStdlib-apinotes
+add_custom_command_target(unused_var
+    CUSTOM_TARGET_NAME CxxStdlib-apinotes
+    OUTPUT "${output_dir}/std.apinotes"
```
This should be after `COMMAND` lines

rintaro committed:
[CMake] Update source file list only if different

Write source file list using 'file(GENERATE' so that they are updated
only if thier content is changed. Otherwise stdlib modules are rebuild
everytime cmake configuration happens.

rintaro committed:
[CMake] Copy '.private.swiftinteface' to swift_static.

'.private.swiftinterface' should be a part of the module. And it's
declared as `OUTPUT` file list, but they were never created. Because of
that the target was always rebuilt.

rintaro commented:
@swift-ci Please smoke test

bnbarham commented on stdlib/cmake/modules/SwiftSource.cmake:
```diff
@@ -921,7 +921,10 @@ function(_compile_swift_files
   string(SHA1 file_name "'${source_files_quoted}'")
   set(file_path "${CMAKE_CURRENT_BINARY_DIR}/${file_name}.txt")
-  file(WRITE "${file_path}" "'${source_files_quoted}'")
+  file(WRITE "${file_path}.tmp" "'${source_files_quoted}'")
+  add_custom_command(
+    OUTPUT "${file_path}"
+    COMMAND ${CMAKE_COMMAND} -E copy_if_different "${file_path}.tmp" "${file_path}")
```
Could we just use `GENERATE` here?
```
file(GENERATE OUTPUT "${file_path}"
     CONTENT "'${source_files_quoted}'")
```

rintaro replied:
Nice! I didn't know `file(GENERATE`.

rintaro commented:
@swift-ci Please smoke test

bnbarham reviewed (approved):
Nice!

compnerd commented on stdlib/cmake/modules/SwiftSource.cmake:
```diff
@@ -1001,7 +1001,9 @@ function(_compile_swift_files
       set(command_copy_interface_file)
       if(interface_file)
-        set(command_copy_interface_file COMMAND "${CMAKE_COMMAND}" "-E" "copy" ${interface_file} ${interface_file_static})
+        set(command_copy_interface_file
```
@DougGregor and I spoke about this - we need to build things in two forms, not just copy the swift interfaces.

compnerd reviewed (changes requested):
Please drop the static swift interfaces or build the libraries appropriately.

bnbarham commented:
> Please drop the static swift interfaces or build the libraries appropriately.

Does that really need to be fixed in this PR? This is just fixing up rebuilds.

rintaro committed:
[CMake] Update source file list only if different

Use 'copy_if_different' instead of 'file(GENERATE'. Somehow
'file(GENERATE' failed with

  CMake Error: Files to be generated by multiple different commands:

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
```
CMake Error: Files to be generated by multiple different commands: "/Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/build/buildbot_incremental/swift-macosx-x86_64/stdlib/public/Cxx/fba1e895a41a8251cfd70af7b678bc1b1e2ae381.txt"
```
This was because the function is called for each stdlib platform/arch , but the file name doesn't take those into account. So use `copy_if_different` instead.

rintaro commented:
> Please drop the static swift interfaces or build the libraries appropriately.

This is really not in the scope of this PR. This PR is only fixing the build issue. If the copying `.swiftinterface` is a problem, that should be fixed separately.
